### PR TITLE
Moving image data to the image

### PR DESCRIPTION
**Pull request by doofus-01**
This is the child of https://github.com/wesnoth/wesnoth/issues/8863

Figuring out how to do this without causing a lot of headaches isn't trivial - I will not address the actual image EXIF tag changes (all the images) here.  The CI will also have to wait.  For now, it's just getting these two maintenance scripts (derived from the existing `update_copyrights` script) to be ready.

Ideally, the workflow goes like this:
1.  Run csv2img-exif to read the data from the existing `copyrights.csv` and write that to the Exif tags of the wesnoth images
2.  Run img-exif2csv to read back the Exif tag data and give a summary of the Exif tag situation.
3.  Any new image in a PR is checked by CI for the needed tags
4.  Periodically, someone runs these two scripts, in case CI didn't catch something.

The initial run of `img-exif2csv` yields the attached CSV.
[img-exif2csv_output-original.csv](https://github.com/user-attachments/files/15942359/img-exif2csv_output-original.csv)

`csv2img-exif` got hung up on some weird data in EI, so this isn't ready for prime-time.

### CODE REVIEW

**Comment by doofus-01** [comment]
The scripts now fully function as intended.  They do take a while to run, there is probably something that should be changed, but they _should_ only need to be run once in a blue moon.  

The image output is what really matters, but what I can post here are the output CSV of these two scripts.
[img-exif2csv_output.csv](https://github.com/user-attachments/files/15948154/img-exif2csv_output.csv)
[csv2img-exif_output.csv](https://github.com/user-attachments/files/15948155/csv2img-exif_output.csv)
I'd like to proceed like this:
1. Merge these two files in this PR (low risk)
2. Merge the image file updates in a subsequent PR (high risk, but is the real meat).  Incidentally, the `.webp` file sizes are significantly smaller after this, due to `cwebp`.
3. Update the CI workflow in a third PR (if that's the way to do it?) to check that pushed images have the correct Exif tags (medium risk, but difficult to test)

**Comment by doofus-01** [comment]
I requested Pentarctagon's review because it is derived from and functionally related to his copyrights script.

**Comment by Pentarctagon** [comment]
Should this still be marked as a draft?

**Comment by Pentarctagon** [comment]
> Merge the image file updates in a subsequent PR (high risk, but is the real meat). Incidentally, the .webp file sizes are significantly smaller after this, due to cwebp.

While I know it's not part of this PR, I'm a bit concerned that the file sizes of existing webp images dropped significantly. This might be due to using lossy mode instead of lossless mode on webp images previously saved as being lossless?

File: csv2img-exif
Comment (soliton-) [inline csv2img-exif:53]: As `utils/optiwebp.py` this should use `-lossless` and `-exact` and probably all other options used there as well.

https://github.com/wesnoth/wesnoth/blob/e67a91caf22429c56cd2012e324cd8ee212b5481/utils/optiwebp.py#L93

Comment (doofus-01) [reply csv2img-exif:53]: why?

**Comment by doofus-01** [comment]
>This might be due to using lossy mode instead of lossless mode on webp images previously saved as being lossless?

Down-scaling and throwing out the color profiles are lossier than anything done with lossy compression, seems that ship has sailed long ago.

**Comment by Pentarctagon** [comment]
> Down-scaling and throwing out the color profiles are lossier than anything done with lossy compression, seems that ship has sailed long ago.

Probably a dumb question, but what is this referring to?

**Comment by doofus-01** [comment]
It's not a dumb question, color profiles are just something we mostly don't have to think about because there are standard defaults (sRGB).  I think it's most important for print magazine photography, but if you are using color management, tossing out the embedded data would distort the colors from the artist's intent.  Not very important most of the time, just like "lossless", but defining color profiles is an option in GIMP (and I assume Photoshop and others).  The folks responsible for this can explain it better - https://www.color.org/faqs.xalter  

More to the point here, I had thought lossy was the only way to make this metadata plan work, now I see I was wrong so I care much less.  I don't think lossless, especially as practiced in `optiwebp.py` was all that well thought out, but I'm not trying to hyper-optimize anything, so lossless it is.

**Comment by Pentarctagon** [comment]
The thought behind optiwebp was essentially just "if jpg then it's already lossy, so it's fine to use lossy webp. if png then it's lossless and should be kept that way". In particular since the original plan was to convert almost everything into webp and you can't have unit sprites saved lossily since that'll mess with the team coloring.

I'll try to test these scripts out locally and review them either tomorrow or the next day (aka Sunday or Monday).

**Comment by Pentarctagon** [comment]
Tried running this locally and saw this warning:
```
Warning: ICC_Profile deleted. Image colors may be affected - doc/manual/images/de/multiplayer-1.14.8+dev.png
Warning: ICC_Profile deleted. Image colors may be affected - doc/manual/images/de/game-screen-1.14.8+dev.jpg
Warning: ICC_Profile deleted. Image colors may be affected - doc/manual/images/de/main-menu-1.14.8+dev.jpg
Warning: ICC_Profile deleted. Image colors may be affected - doc/manual/images/de/recruit-1.14.8+dev.png
Warning: ICC_Profile deleted. Image colors may be affected - doc/manual/images/de/right-pane-1.14.8+dev.png
```
Is there any concern with that?

It also then failed with the error:
> Error: Writing of WEBP files is not yet supported - data/core/images/maps/background.webp

What linux distro/version are you using?

**Comment by doofus-01** [comment]
The ICC profiles were already deleted from most of the other images, I didn't see any reason to think those were intentionally being saved.

I'm currently using exiftool version 12.87 because I wanted to see if it could write RIFF tags for `.webp` (still no), but the years-older version that came with openSUSE Leap 15.5 was able to handle `.webp` (I don't remember the version).  Asking a maintainer to manually install an external tool isn't great, but I remember `utils/woptipng.py` couldn't work for me with just my distribution's repositories; exiftool is a relatively small installation.  Maybe a "minimum tested version" comment would help?

**Comment by Pentarctagon** [comment]
Looks like it requires at least version 12.46 (https://exiftool.org/forum/index.php?topic=13797.msg75297#msg75297) and I only have version 12.40 locally. The version of Mint I have is based on the same version of Ubuntu that the CI uses though, so they should have the same version of software. If you've run it on the CI separately from this PR and it works though, I guess if it works then it works.

**Comment by soliton-** [comment]
I don't see exiftool in the github runner images or our docker images. Ubuntu 24.04 does have exiftool 12.76 though so a future docker image could have that available.

**Comment by doofus-01** [comment]
This was my CI test, it would be a starting-point for the future PR.
https://github.com/doofus-01/wesnoth-Bad_Moon_Rising/blob/master/.github/workflows/check_metadata.yml

**Comment by Pentarctagon** [comment]
Huh, that's really weird. The version installed from the last CI run there shows that it matches what I have locally, so no idea why it supports webp on CI but not on my local PC.

**Comment by CelticMinstrel** [comment]
Maybe webp support is optional in the build, and yours was built without it? If it was installed via a package manager, you could try seeing if there's a "with webp" variant.

**Comment by Pentarctagon** [comment]
The CI run is https://github.com/doofus-01/wesnoth-Bad_Moon_Rising/actions/runs/9246349622/job/25433792199, which also is just installing it from the package manager:
`sudo apt-get install --assume-yes exiftool`

It also installs apparently the same version `12.40+dfsg-1`.

**Comment by doofus-01** [comment]
It could just be that v12.40 can read the data from `.webp` but can't write it.  That's currently how it is with RIFF tags.

**Comment by doofus-01** [comment]
@soliton- , @Elvish-Hunter - Do you have any further thoughts on this?
Sorry for being pushy, but this will will change the image workflow, it would be great to not let this linger.
The performance can always be improved later, as long as the image output is acceptable.

**Comment by Elvish-Hunter** [comment]
Sorry, I completely missed this.
Anyway, the scripts use two command line tools to do their job: exiftool and cwebp. I was wondering if it would make sense to use the Pillow library to handle the EXIF tags instead of exiftool.

**Comment by Pentarctagon** [comment]
I downloaded 12.78 from the exiftool website and was able to use it. Both scripts do definitely take quite a while to run, but their output looks good (though obviously I didn't check every image).

**Review by Pentarctagon — APPROVED** [review]

**Comment by soliton-** [comment]
I'm not a fan of littering our repository root more. IMO the end result should be a single script to check if all images have the required metadata. The script could have an option to produce the `copyrights.csv` if that is still useful for something but I don't think it makes sense to have images with metadata and `copyrights.csv` checked in at the same time. That just invites inconsistencies. (For a transition period that's of course not a big deal though.)

The single script could live in the repo root if that's really important but I'd rather put it under utils.

**Comment by doofus-01** [comment]
> I was wondering if it would make sense to use the Pillow library to handle the EXIF tags instead of exiftool.

The output shouldn't depend on the tool, so if you're ready to obsolete this before it's used, I won't object.

> I'm not a fan of littering our repository root more. IMO the end result should be a single script to check if all images have the required metadata. The script could have an option to produce the copyrights.csv if that is still useful for something but I don't think it makes sense to have images with metadata and copyrights.csv checked in at the same time. That just invites inconsistencies. (For a transition period that's of course not a big deal though.)

Producing the CSV is just so there is something easy to browse afterwards, the CSV outputs probably shouldn't even be added to the repository.  These scripts are pretty much for transition period anyway, or at least the `csv2img-exif` is.

>The single script could live in the repo root if that's really important but I'd rather put it under utils.

No, not important, that's just where the original copyrights script was.  I've moved it to utils.

**Comment by soliton-** [comment]
IMO this looks good. (I've not looked closely at the python code.)

Regarding adding the tags on future images do we expect artists will manage to do that on their own? Or should we provide a script to make it easier? Or the artist provides the info and some dev (maybe even the CI) does it?

Can be all of the above of course.

**Review by Elvish-Hunter — APPROVED** [review]

File: utils/csv2img-exif
Comment (soliton-) [inline utils/csv2img-exif:130]: This should at a minimum leave fields 3 and 4 empty or the check for empty further down will not consider these entries as incomplete.

**Comment by soliton-** [comment]
I think the default of `--repo ..` is not great. Besides being broken here and in `update_copyrights` i'd expect people to run the script from the repo root by default.

File: utils/csv2img-exif
Comment (soliton-) [inline utils/csv2img-exif:103]: Since the current directory was already changed to `options.repo` this should just go through the current dir. Otherwise this goes wrong for relative paths.

File: utils/csv2img-exif
Comment (soliton-) [inline utils/csv2img-exif:108]: This is probably a workaround for the broken relative repo path.

**Review by soliton- — CHANGES_REQUESTED** [review]

**Comment by CelticMinstrel** [comment]
> I think the default of `--repo ..` is not great. Besides being broken here and in `update_copyrights` i'd expect people to run the script from the repo root by default.

I think it should run on `.` by default. The expectation should be that it's run as `./utils/script-name`, rather than `cd utils; ./script-name`.

**Comment by doofus-01** [comment]
I've reverted the `../utils` stuff, just moved the scripts there.  Did I understand the request correctly?

**Comment by CelticMinstrel** [comment]
I think so?

**Comment by Pentarctagon** [comment]
I mean, that's a yes or no question.

**Comment by Pentarctagon** [comment]
Well, I'll just say then that yes, it looks like what was requested.

**Comment by doofus-01** [comment]
Thanks, Pentarctagon.  I need to figure out if there is anything I should include regarding AI tagging, but it's looking like that is a future issue.  Otherwise, I'll commit these next weekend.

File: utils/csv2img-exif
Comment (soliton-) [inline utils/csv2img-exif:110]:
```suggestion
                added.append(["NEW", file, "", "", "", "NO_DATA", "NO_DATA"])
```
IMO something like "NO_DATA" is of questionable use and should definitely not be used when "NO_DATA" is valid data itself.

File: utils/csv2img-exif
Comment (soliton-) [inline utils/csv2img-exif:128]:
```suggestion
                added.append(["NEW", file, "", "", "", "NO_DATA", "NO_DATA"])
```

File: utils/csv2img-exif
Comment (soliton-) [reply utils/csv2img-exif:103]:
```suggestion
for root, _, files in os.walk("."):
```

File: utils/csv2img-exif
Comment (soliton-) [inline utils/csv2img-exif:147]:
```suggestion
    with open(options.output, 'w', encoding="utf-8") as f:
```
As per #9098.

File: utils/img-exif2csv
Comment (soliton-) [inline utils/img-exif2csv:54]:
```suggestion
with open(options.output, 'w', encoding="utf-8") as f:
```

File: utils/img-exif2csv
Comment (soliton-) [inline utils/img-exif2csv:56]:
```suggestion
  for root, _, files in os.walk("."):
```

**Review by soliton- — CHANGES_REQUESTED** [review]

**Comment by soliton-** [comment]
> I need to figure out if there is anything I should include regarding AI tagging, but it's looking like that is a future issue.

You mean how to tag files that were generated solely or with the help of AI? I'd leave that up to the point where we change policy on AI art.

**Comment by doofus-01** [comment]
> You mean how to tag files that were generated solely or with the help of AI? I'd leave that up to the point where we change policy on AI art.

No, I meant if there should be some tag for an AI training opt-out (not that it should be opt-out).  I didn't want to have to touch every image again to add whatever tag that would be, whatever the consensus was on the default.  But regardless of whether Wesnoth images should be scraped for AI or not, it doesn't look like there is anything special to add to the images, aside from the author and copyright.  

There is a "NoAI" html tag, but that would be more for the website, like `robots.txt`, and since most AI companies won't honor it, it's probably just as well to have a policy statement somewhere.

So this is basically a non-issue, I just had to check.

File: utils/csv2img-exif
Comment (doofus-01) [reply utils/csv2img-exif:110]: "NO_DATA" was just meant as some easy-to-search handle that came from this script, but it isn't critical.  "NEW" for the file path sounds even more questionable.

**Comment by Pentarctagon** [comment]
We could add those as well to tell AI crawlers to skip the wesnoth site, but yeah I'm not sure how much they actually honor that sort of thing.

File: csv2img-exif
Comment (soliton-) [reply csv2img-exif]: Doesn't look like this got implemented properly yet. The image data should not be changed, right? So we should use `--lossless` and `--exact`.

Comment (soliton-) [reply csv2img-exif]: Actually what is the `cwebp` call even for?

**Comment by soliton-** [comment]
If there is some consensus on what such tags should look like then we could certainly add them. Looks like there is http://iptc.org/std/photometadata/specification/IPTC-PhotoMetadata-2023.1.html#data-mining at least.

**Comment by doofus-01** [comment]
I haven't found anyone using that yet (I could easily be missing something), but that certainly does seem on target if it becomes an accepted standard.  I'm not sure any tag will have more meaning than a general AI policy post at wesnoth.org, but probably best to move this discussion to the forums.

I'll consider the AI tag out of scope here, since there are still issues to discuss on that.  If there are no other issues, I'll merge this next weekend, and then use these to update the images in a new PR.  The current copyrights tracking can't be retired until the CI check is implemented, and that's the part that makes me nervous.

**Comment by soliton-** [comment]
I've asked what the cwebp call is for above. (Easy to miss since I also said something "down here" and multiple notifications for a PR are not handled well by github.)

File: csv2img-exif
Comment (doofus-01) [reply csv2img-exif]: > Doesn't look like this got implemented properly yet. The image data should not be changed, right? So we should use --lossless and --exact.

It's properly implemented, aside from the typo in "-mt".  Lossless is implied by "-z" and "-exact" only preserves the invisible garbage.

> Actually what is the cwebp call even for?

Exiftool can't yet add RIFF alpha tags to WEBP, so it produces an Extended WEBP without an Alpha flag.

**Comment by doofus-01** [comment]
Sorry, I did miss it earlier but I've addressed it now.

**Comment by doofus-01** [comment]
This seems to be working, and after just having had to update copyrights.csv for 70 images in the cats PR, I'd like to keep this moving.  So I will merge next weekend, assuming there are no further objections.

The music files already have Vorbis comments, so I'd rather not mess with those.  I guess the sound files could be included here, but that will require a different CLI tool, no real reason it has to be done with the images.  The sound files are a much smaller set than images, so manual tracking during the transition isn't as much of an issue.

File: csv2img-exif
Comment (soliton-) [reply csv2img-exif]: I don't really understand why we need different options here but I suppose it shouldn't make a difference.

It would be good to explain the use of cwebp in a comment so that we can tell once it's not needed anymore.

Comment (doofus-01) [reply csv2img-exif]: I've added a comment

**Review by soliton- — APPROVED** [review]

**Pull request merged**
